### PR TITLE
Delete s3aio driver

### LINKED ISSUE #863: Remove s3aio drivers
`s3aio` driver was an experiment in using `zarr`-like storage on S3, before `zarr` became a well known thing.

My understanding is that it is no longer being used or developed. Nor is it a recommended way for storing/accessing imagery on S3. 

Recommended way is to use Cloud Optimized GeoTIFFs on S3, this is done by the default, rasterio based driver. It can be confusing to new users who want to utilise S3 with datacube, as they assume that s3aio is the way to go.

There is also non-trivial amount of developing overhead, mostly to do with tests and environment setup.

### PR DESCRIPTION
**Pull request by Kirill888**
### Reason for this pull request

S3AIO driver served it's purpose as an experiment. We will consider using regularly grided data backed by zarr in the future as part of planned 2.0 work. Removing s3aio as it is not being actively developed or used to ease maintenance burden. 

### Proposed changes

- Remove S3AIO driver code, tests and documentation

 - [x] Closes #863 
 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

### CODE REVIEW

**Comment by petewa** [comment]
Please don't remove S3AIO, this is a required component for our project and work should begin in April to add support for 4D hyperspectral/lidar data with a Zarr backend.

**Comment by Kirill888** [comment]
@petewa you should be able to use it as a stand-alone component, this was the whole point of us developing plugin infrastructure. We had a work-around with `_legacy` a year ago, yet no work was done to remove that workaround in the meantime. We can not afford to continue to maintain this without any support from original developers.

**Comment by petewa** [comment]
@Kirill888 I/we would have loved to work on S3AIO earlier, but unfortunately it didn't happen.
We intend to start work in April to extend S3AIO for 4D Hyperspectral/Lidar data with a Zarr backend.

**Comment by Kirill888** [comment]
@petewa it's great that you'll get time to work on this soon, but may I suggest that you use provided plugin infrastructure when you start, and develop it in a separate repository. This was always the original goal back 2+ years ago when plugin code was written.

As it stands right now, we here in GA, are stuck with maintenance of non-trivial amount of code we don't understand nor use, so any failures that are not covered by tests, and test coverage in that section is relatively low, will cause failures. 

Other aspect of this is external user confusion, people see S3 in the name and assume that this is the way to use S3 with datacube and it is not the case. This also creates support load on Slack.

**Comment by Kirill888** [comment]
For an example on how to hook in your code into datacube plugin infrastructure see here:

https://github.com/opendatacube/datacube-core/blob/e8d0a0c4c8f24d0251c6222c215f10a02d2aad16/tests/drivers/fail_drivers/setup.py#L1-L17

**Comment by mikec-codes** [comment]
The s3aio driver is required for our hyperspectral project ....by deprecating it will cause significant delays to our project  .  I suggest we hold off deprecating the driver.

**Comment by Kirill888** [comment]
@m7200  and @petewa how about this:

We retain legacy hook-in, but nothing else. You maintain s3aio in a separate repository, it's on you to port tests over to your setup and to run and monitor and react to them.

Legacy callback will not be moved to the next version.

**Review by woodcockr — CHANGES_REQUESTED** [review]
Need to put a short hold on this PR so another project can prepare appropriately.
Should have some resolution late next week as we're meeting face to face.

**Comment by Kirill888** [comment]
Well it's the end of the week, any resolutions? @woodcockr @petewa @m7200

**Comment by woodcockr** [comment]
@Kirill888 Workshop finishes 5pm, so not quite the end of the week. ;-)

**Comment by woodcockr** [comment]
Reporting back with my ODC SC Chair hat on - the CSIRO project team associated with this change has discussed the refactor process for s3aio and the ODC 2 path for this project with the associated Project Leaders and Stakeholders today. There is general understanding of the proposed PR and the reasons associated with it (maintenance, integration testing impacts). The CSIRO team does have ongoing support now for the development of an updated version of ODC to support hyperspectral data per the improvement proposals ODC SC has for multi-dimensional support planned for v2.0.
There is desire to preserve some components of the s3AIO code base, particularly the test harness for multidimensional support and in order to do this to refactor the component to use the plugin approach discussed by @Kirill888 earlier in this thread.
There are some details to be worked out so they can be put back to the ODC SC for consideration and the developers from project team intends on meeting again next week to sort these details next week now the project stakeholders have been consulted

**Review by woodcockr — APPROVED** [review]

**Comment by woodcockr** [comment]
@Kirill888 at al. Took a little longer than I hoped but as ODC SC Chair I've been in touch with impacted CSIRO project and can now approve this PR.
There will be an update down the track to the [Add support for 3D datasets](https://github.com/opendatacube/datacube-core/wiki/ODC-EP-001---Add-Support-for-3D-Datasets) proposal and associated design discussions the ODC community has been having. The CSIRO project seems keen on adding a Zarr backend using the plugin support for ODC 1.x series and contributing to the ODC 2.0 implementation. Please note I am not making a commitment on their behalf with this, just indicating that there was more to the discussion than just deprecating s3aio at this end.

**Comment by woodcockr** [comment]
I neglected to mention, please ask one of your core developer colleagues to do the peer review. I'm not in a position to verify the PR works beyond eyeballing the changes (which I did) just at the moment - preparing for a big trip next week!

File: docs/architecture/driver.rst
Comment (omad) [inline docs/architecture/driver.rst:192]: This is still relevant I think

**Review by omad — CHANGES_REQUESTED** [review]
Looks pretty safe to me, and the tests pass, just one query about the docs.

File: docs/architecture/driver.rst
Comment (Kirill888) [reply docs/architecture/driver.rst:192]: yep, will resurrect

**Review by omad — APPROVED** [review]
Thank you. :)

**Pull request merged**
